### PR TITLE
Update websocket relay example app

### PR DESCRIPTION
**Pull request by bgins**
# Description

This PR implements the following changes:

- [x] Update example to use Homestar client: https://www.npmjs.com/package/@fission-codes/homestar
- [x] Remove emulation mode

## Link to issue

Implements #390

## Type of change

- [x] Refactor (non-breaking change that updates existing functionality)

## Test plan (required)

The example should be tested manually to verify it works.

### CODE REVIEW

**Review by avivash — APPROVED** [review]
LGTM! One small thing: there are a couple low severity npm audit issues that can be resolved with `npm audit fix`

image

**Comment by bgins** [comment]
> there are a couple low severity npm audit issues that can be resolved with `npm audit fix`

Ah yeah, good call! Updated. 👌

**Review by hugomrdias — CHANGES_REQUESTED** [review]

File: examples/websocket-relay/relay-app/package.json [hunk 1]
```diff
@@ -37,8 +37,11 @@
   },
   "type": "module",
   "dependencies": {
+    "@fission-codes/homestar": "^0.6.2",
     "@zerodevx/svelte-json-view": "^1.0.3",
     "clipboard-copy": "^4.0.1",
+    "iso-base": "^2.0.1",
+    "multiformats": "^12.1.3",
```
[the remark this answers is not in the document]
Comment (bgins) [reply examples/websocket-relay/relay-app/package.json:44]: `multiformats` is used for `CID` when constructing workflows, for example:

https://github.com/ipvm-wg/homestar/blob/73f7955c70a59da43ac936a0e1518d785264eb25/examples/websocket-relay/relay-app/src/lib/workflow.ts#L265-L267

This field reported a type error when attempting to use a string.

`iso-base` is used for converting the `Uint8array` in a receipt to base64:

https://github.com/ipvm-wg/homestar/blob/73f7955c70a59da43ac936a0e1518d785264eb25/examples/websocket-relay/relay-app/src/lib/workflow.ts#L233

**Comment by hugomrdias** [comment]
Where they just missing in package.json I don't see new imports for them in
this PR

Hugo Dias

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 2]
```diff
@@ -1,16 +1,18 @@
+import { CID } from "multiformats";
+import { base64 } from "iso-base/rfc4648";
```
Comment (bgins) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:2]: @hugomrdias the imports for `multiformats` and `iso-base` are here.

Comment (hugomrdias) [reply examples/websocket-relay/relay-app/src/lib/workflow.ts:2]: Got that file shows collapsed and seemed to be just removed lines. I will review further. 
Ty

Comment (bgins) [reply examples/websocket-relay/relay-app/src/lib/workflow.ts:2]: Thanks for looking over it! Would be nice to not need `multiformats` here if it's possible to use a string.

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 3]
```diff
@@ -1,16 +1,18 @@
+import { CID } from "multiformats";
```
Comment (hugomrdias) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:1]:
```suggestion
import { CID } from "multiformats/cid";
```

for now i really need the CID instance to be able to serialize tasks as dag-cbor

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 4]
```diff
@@ -132,90 +158,83 @@ export function fail(workflowId: WorkflowId) {
 
 // HANDLER
 
-export async function handleMessage(event: MessageEvent) {
-  const data = await event.data.text();
-
+export async function handleMessage(data: MaybeResult) {
```
Comment (hugomrdias) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:161]: this should be `MaybeResult<Schemas.WorkflowNotification, Schemas.WorkflowNotificationError>` so you get the proper types all the way down i will export more types from the entry point and release a new version

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 5]
```diff
@@ -132,90 +158,83 @@ export function fail(workflowId: WorkflowId) {
...
-function parseReceipt(raw: {
+type RawReceipt = {
   iss: string | null;
   meta: Meta | null;
-  out: ["ok" | "error", Record<"/", Record<"bytes", string>>];
+  out: ["ok" | "error", Uint8Array];
   prf: string[];
   ran: Record<"/", string>;
-}): Receipt {
+};
```
Comment (hugomrdias) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:227]: i will export this type so you can remove this

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 6]
```diff
@@ -132,90 +158,83 @@ export function fail(workflowId: WorkflowId) {
...
+
+const parseReceipt = (raw: RawReceipt): Receipt => {
   return {
     iss: raw.iss,
     meta: raw.meta,
-    out: [raw.out[0], raw.out[1]["/"].bytes],
+    out: [raw.out[0], base64.encode(raw.out[1])],
```
Comment (hugomrdias) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:233]: why are you base encoding this ? its ui related?

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 7]
```diff
@@ -233,226 +252,75 @@ function getTaskMessage(status: TaskStatus) {
...
+export const workflowOnePromised = WorkflowBuilder.workflow({
+  name: "one",
+  workflow: {
+    tasks: [
+      WorkflowBuilder.crop({
+        name: "crop",
+        resource: "bafybeiczefaiu7464ehupezpzulnti5jvcwnvdalqrdliugnnwcdz6ljia",
+        args: {
+          data: CID.parse(
+            "bafybeiejevluvtoevgk66plh5t6xiy3ikyuuxg3vgofuvpeckb6eadresm"
```
Comment (hugomrdias) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:266]:
```suggestion
          data: "{{ cid:bafybeiejevluvtoevgk66plh5t6xiy3ikyuuxg3vgofuvpeckb6eadresm }}"
```
actually you can do this and avoid `CID` dependency

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 8]
```diff
@@ -61,27 +58,56 @@ export async function run(workflowId: WorkflowId) {
...
+            ? { ...store, failedPingCount: activeWorkflow.failedPingCount + 1 }
+            : null
+        );
+      }
+    } else {
+      // No workflow active
+      clearInterval(interval);
+    }
+  }, import.meta.env.VITE_PING_INTERVAL);
 }
```
Comment (hugomrdias) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:111]: we can add this feature to the client itself.
you should actually subscribe to the client `error` events to handle unexpected behaviour

**Review by hugomrdias — CHANGES_REQUESTED** [review]
left some notes inline, will do some improvements to the client to help here and release a new version

File: examples/websocket-relay/relay-app/src/lib/workflow.ts (hunk 8 above)
Comment (bgins) [reply examples/websocket-relay/relay-app/src/lib/workflow.ts:111]: Yeah! That would be great. 💯

File: examples/websocket-relay/relay-app/src/lib/workflow.ts (hunk 6 above)
Comment (bgins) [reply examples/websocket-relay/relay-app/src/lib/workflow.ts:233]: Yes, we are using it to display the images here:

https://github.com/ipvm-wg/homestar/blob/ebc8fb1595ed01eb7123d69d6503bde2dd98ac3a/examples/websocket-relay/relay-app/src/components/node/Node.svelte#L59

File: examples/websocket-relay/relay-app/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (bgins) [reply examples/websocket-relay/relay-app/package.json]: We were able to remove `CID.parse` and multiformats by wrapping the data CIDs in `{{ ... }}`, for example `{{ cid:bafybeiejevluvtoevgk66plh5t6xiy3ikyuuxg3vgofuvpeckb6eadresm }}`

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 9]
```diff
@@ -132,90 +158,83 @@ export function fail(workflowId: WorkflowId) {
...
+export async function handleMessage(data: MaybeResult) {
   console.log("Received message from server: ", data);
 
-  // Reset ping count on echoed ping or pong from server
-  if (data === "ping" || data === "pong") {
-    activeWorkflowStore.update((store) =>
-      store ? { ...store, failedPingCount: 0 } : null
-    );
+  if (data.error) {
+    // @ts-ignore-next-line
```
[the remark this answers is not in the document]
Comment (bgins) [reply examples/websocket-relay/relay-app/src/lib/workflow.ts]: Done. 👌

File: examples/websocket-relay/relay-app/src/lib/workflow.ts (hunk 4 above)
Comment (bgins) [reply examples/websocket-relay/relay-app/src/lib/workflow.ts]: Updated with the switch to `@fission-codes/homestar` 1.0.0.

File: examples/websocket-relay/relay-app/src/lib/workflow.ts [hunk 10]
```diff
@@ -132,90 +162,75 @@ export function fail(workflowId: WorkflowId) {
...
 
-  // Reset ping count on echoed ping or pong from server
-  if (data === "ping" || data === "pong") {
-    activeWorkflowStore.update((store) =>
-      store ? { ...store, failedPingCount: 0 } : null
-    );
+  if (data.error) {
+    // @ts-expect-error received WorkflowNotificationError from Homestar node
+    throw new Error(data.error);
+  }
```
Comment (hugomrdias) [inline examples/websocket-relay/relay-app/src/lib/workflow.ts:173]: what are we ignoring here? you can probably throw `data.error` directly or `new Error('bad message', { cause: data.error })`

**Review by hugomrdias — CHANGES_REQUESTED** [review]

File: examples/websocket-relay/relay-app/src/lib/workflow.ts (hunk 10 above)
Comment (bgins) [reply examples/websocket-relay/relay-app/src/lib/workflow.ts]: You're right, nothing left to ignore. Updated to remove the `@ts-expect-error` and throw `data.error` directly.

**Review by hugomrdias — APPROVED** [review]

**Comment by zeeshanlakhani** [comment]
@bgins this all looks great. One minor change maybe. Since we have the option to not pass settings anymore, and we don't really use it in the example, can we update to go that direction for the example (as easy as possible?)?

**Review by zeeshanlakhani — APPROVED** [review]
Generally approved, but had one minor ask.

**Comment by bgins** [comment]
> @bgins this all looks great. One minor change maybe. Since we have the option to not pass settings anymore, and we don't really use it in the example, can we update to go that direction for the example (as easy as possible?)?

Yeah, good idea! Added that.

**Pull request merged**